packages/netlify-cms-backend-gitlab/src/API.js

```diff
@@ -224,58 +260,449 @@ export default class API {
+  async uploadAndCommit(
+    entry,
+    items,
+    { commitMessage, newEntry = true, branch = this.branch, author = this.commitAuthor },
+  ) {
+    const commitParams = this.toCommitParams(commitMessage, branch, author);
+    const action = newEntry ? 'create' : 'update';
+    const actions = await Promise.all(
+      items.map(async item => ({
+        action,
+        file_path: item.path.replace(/^\//, ''),
+        content: await result(item, 'toBase64', partial(this.toBase64, item.raw)),
+        encoding: 'base64',
+        ...(item.sha ? { last_commit_id: item.sha } : {}),
+      })),
+    );
+    const response = await this.requestJSON({
       method: 'POST',
-      headers: { 'Content-Type': 'application/json' },
-      body: JSON.stringify(commitParams),
+      url: `${this.repoURL}/repository/commits`,
+      body: {
+        ...commitParams,
+        actions,
+      },
     });
+    return items.map(item => ({
+      ...item,
+      sha: response.id,
+      uploaded: true,
+    }));
+  }
 
-    return { ...item, uploaded: true };
-  };
+  persistFiles(entry, mediaFiles, options) {
+    const files = entry ? [entry, ...mediaFiles] : [...mediaFiles];
+    if (!options.useWorkflow) {
+      return this.uploadAndCommit(entry, files, options);
+    } else {
+      return this.editorialWorkflowGit(entry, files, options);
+    }
+  }
+
+  deleteFile(path, commitMessage, { branch = this.branch, author = this.commitAuthor }) {
+    const commitParams = this.toCommitParams(commitMessage, branch, author);
+    return this.request({
+      method: 'DELETE',
+      url: `${this.repoURL}/repository/files/${encodeURIComponent(path)}`,
+      params: commitParams,
+      // Note that changes to the file are allowed, so `last_commit_id` is not checked against `item.sha`.
+      // last_commit_id: ???,
+    });
+  }
+
+  // Metadata
+
+  checkMetadataBranch() {
+    return this.getBranch(CMS_METADATA_BRANCH).catch(error => {
+      if (error instanceof APIError && error.status === 404) {
+        return this.createMetadataBranch();
+      }
+      throw error;
+    });
+  }
+
+  createMetadataBranch() {
+    // The Branches API doesn't support creating orphan branches.
+    // The Commits API doesn't support creating orphan branches.
+    //   https://gitlab.com/gitlab-org/gitlab-ce/issues/2420 only works for empty repositories.
+    throw new Error(`Not Implemented - GitLab API does not support creating orphan branches.'`);
+    // TODO - rudolf - Update README / Installation and Configuration.
+    // Git could be used for creating orphan branches.
+    //   git checkout --orphan _netlify_cms
+    //   git rm --cached *
+    //   echo "# Netlify CMS\n\nThis branch is used by Netlify CMS to store metadata information for files, branches, merge requests.\n" > README.md
+    //   git add README.md
+    //   git commit -m "Initial commit by Netlify CMS"
+    //   git push --set-upstream origin _netlify_cms
```

Comment:
Why it's so important to create an orphan branch?

I mean it looks cleaner okay, but if this is not supported by GitLab, I don't see any problem to create just a normal branch or based on the first commit in the repo.

Reply:
Definitely, we're aware that orphan refs are GitHub only - we're actually moving away from this model entirely and using more obvious methods like PR/MR labels to track status.

Here's an issue outlining that: https://github.com/netlify/netlify-cms/issues/1669

And here's a WIP PR implementing it for GitHub: https://github.com/netlify/netlify-cms/pull/1961